File: aisearch/rpc.py
"""
stdio 无服务模式 —— 不需要监听任何端口。

用法（AI 集成推荐）：
    aisearch rpc                # 以 cwd 为项目根，CLI 级信任（boundary=system）
    aisearch rpc --root /path   # 锁定项目根并限制 boundary=root

协议：行分隔 JSON（stdin 进，stdout 出，每请求一行响应）
    请求: {"id": 1, "method": "search", "params": {"pattern": "def foo"}}
    响应: {"id": 1, "ok": true, "data": {...}}
    错误: {"id": 1, "ok": false, "error": "..."}

方法：search / symbols / def / ref / read / context / tree / health

rpc 由调用方（AI harness）以子进程拉起，
无端口、无网络监听、无鉴权需求；进程退出即结束，适合本地 AI Agent。
"""

from __future__ import annotations

import json
import sys
from pathlib import Path

from . import __version__, engine, reader
from .config import find_project_root, is_relative_to


MAX_LINE_BYTES = 1 * 1024 * 1024

_METHODS = ("search", "symbols", "def", "ref", "read", "context", "tree", "health")


def _clamp(value, default, lo, hi):
    try:
        n = int(value)
    except (TypeError, ValueError, OverflowError):
        return default
    return max(lo, min(hi, n))


class RpcSession:

    def __init__(self, default_path: str = ".", boundary: str = "system"):
        self.default_path = default_path
        self.boundary = boundary



    def handle(self, req: dict) -> dict:
        rid = req.get("id")
        method = req.get("method", "")
        params = req.get("params") or {}
        if not isinstance(params, dict):
            return {"id": rid, "ok": False, "error": "'params' must be an object"}

        handler = getattr(self, f"_m_{method}", None)
        if handler is None:
            return {
                "id": rid, "ok": False,
                "error": f"Unknown method '{method}'; available: {', '.join(_METHODS)}",
            }
        try:
            data = handler(params)
        except Exception as e:
            return {"id": rid, "ok": False, "error": f"{type(e).__name__}: {e}"}
        return {"id": rid, "ok": True, "data": data}



    def _path(self, params: dict) -> str:
        return params.get("path") or self.default_path

    def _check_path(self, params: dict) -> str:
        p = self._path(params)
        if self.boundary == "root":
            base = Path(self.default_path).resolve()
            root = find_project_root(p).resolve()
            if not is_relative_to(root, base):
                raise ValueError(f"Path '{p}' escapes project root")
        return p

    def _m_health(self, params: dict) -> dict:
        return {"version": __version__, "methods": list(_METHODS),
                "root": str(Path(self.default_path).resolve()),
                "boundary": self.boundary}

    def _m_search(self, params: dict) -> dict:
        pattern = params.get("pattern", "")
        if not pattern or not str(pattern).strip():
            raise ValueError("Missing 'pattern'")
        exts = params.get("extensions")
        if isinstance(exts, list):
            exts = {str(e).lower() if str(e).startswith(".") else f".{str(e).lower()}" for e in exts}
        elif isinstance(exts, str):
            exts = {e.lower() if e.startswith(".") else f".{e.lower()}" for e in exts.split(",")}
        resp = engine.search_text(
            pattern=pattern,
            path=self._check_path(params),
            extensions=exts,
            context_lines=params.get("context", 2),
            case_insensitive=bool(params.get("ignore_case", False)),
            max_results=params.get("limit", 50),
            whole_word=bool(params.get("whole_word", False)),
        )
        if not resp.ok:
            raise ValueError(resp.error)
        return resp.to_dict()["data"]

    def _m_symbols(self, params: dict) -> dict:
        name = params.get("name", "")
        if not name:
            raise ValueError("Missing 'name'")
        resp = engine.search_symbols(
            name=name,
            path=self._check_path(params),
            kind=params.get("kind"),
            max_results=params.get("limit", 50),
            partial=bool(params.get("partial", True)),
        )
        return resp.to_dict()["data"]

    def _m_def(self, params: dict) -> dict:
        name = params.get("name", "")
        if not name:
            raise ValueError("Missing 'name'")
        resp = engine.find_definition(
            name,
            self._check_path(params),
            max_results=params.get("limit", 50),
            substring_fallback=bool(params.get("substring_fallback", True)),
        )
        return resp.to_dict()["data"]

    def _m_ref(self, params: dict) -> dict:
        name = params.get("name", "")
        if not name:
            raise ValueError("Missing 'name'")
        resp = engine.find_references(
            name=name,
            path=self._check_path(params),
            max_results=params.get("limit", 100),
        )
        if not resp.ok:
            raise ValueError(resp.error)
        return resp.to_dict()["data"]

    def _m_read(self, params: dict) -> dict:
        file_ref = params.get("file", "")
        if not file_ref:
            raise ValueError("Missing 'file'")
        result = reader.read_file(
            file_ref=file_ref,
            path=self._path(params),
            outline_only=bool(params.get("outline", False)),
            boundary=self.boundary,
        )
        if not result.ok:
            raise ValueError(result.error)
        return result.to_dict()["data"]

    def _m_context(self, params: dict) -> dict:
        file_ref = params.get("file", "")
        line = _clamp(params.get("line", 0), 0, 0, 10**9)
        if not file_ref or line <= 0:
            raise ValueError("Missing 'file' and 'line'")
        result = reader.get_context(
            file_ref=file_ref,
            line=line,
            path=self._path(params),
            radius=_clamp(params.get("radius", 5), 5, 0, 200),
            boundary=self.boundary,
        )
        if not result.ok:
            raise ValueError(result.error)
        return result.to_dict()["data"]

    def _m_tree(self, params: dict) -> dict:
        return engine.project_tree(
            path=self._check_path(params),
            depth=params.get("depth", 3),
        )


def run_rpc(root: str | None = None):


    try:
        sys.stdin.reconfigure(encoding="utf-8", errors="replace")
        sys.stdout.reconfigure(encoding="utf-8")
    except (AttributeError, ValueError):
        pass

    default_path = root if root else "."
    boundary = "root" if root else "system"
    session = RpcSession(default_path=default_path, boundary=boundary)

    stdin = sys.stdin
    stdout = sys.stdout

    def _emit(resp: dict) -> None:

        stdout.write(json.dumps(resp, ensure_ascii=False, separators=(",", ":")) + "\n")
        stdout.flush()

    def _handle_line(line: str) -> None:
        try:
            req = json.loads(line)
            if not isinstance(req, dict):
                raise ValueError("Request must be a JSON object")
        except (json.JSONDecodeError, ValueError) as e:
            _emit({"id": None, "ok": False, "error": f"Invalid request: {e}"})
            return
        _emit(session.handle(req))

    try:




        buf = b""
        dropping = False
        too_large = {"id": None, "ok": False,
                     "error": f"Request line too large (>{MAX_LINE_BYTES} bytes)"}
        while True:
            chunk = stdin.buffer.read1(65536) if hasattr(stdin, "buffer") else stdin.read(65536)
            if not chunk:
                break
            if isinstance(chunk, str):
                chunk = chunk.encode("utf-8", "replace")
            buf += chunk
            while True:
                nl = buf.find(b"\n")
                if nl < 0:
                    if not dropping and len(buf) > MAX_LINE_BYTES:
                        dropping = True
                        buf = b""
                        _emit(too_large)
                    break
                raw, buf = buf[:nl], buf[nl + 1:]
                if dropping:
                    dropping = False
                    continue
                if len(raw) > MAX_LINE_BYTES:
                    _emit(too_large)
                    continue

                line = raw.decode("utf-8", errors="replace").strip().lstrip("\ufeff").strip()
                if not line:
                    continue
                _handle_line(line)
        if buf and not dropping:

            if len(buf) > MAX_LINE_BYTES:
                _emit(too_large)
            else:
                line = buf.decode("utf-8", errors="replace").strip().lstrip("\ufeff").strip()
                if line:
                    _handle_line(line)
    except BrokenPipeError:

        return
    except KeyboardInterrupt:
        return

File: aisearch-js/src/rpc.mjs















import path from "node:path";


import { VERSION, find_project_root, is_relative_to } from "./config.mjs";
import * as engine from "./engine.mjs";
import * as reader from "./reader.mjs";
import { PyValueError, safe_int, abs_path } from "./util.mjs";


export const MAX_LINE_BYTES = 1 * 1024 * 1024;

export const METHODS = ["search", "symbols", "def", "ref", "read", "context", "tree", "health"];


export class RpcSession {
  
  constructor(default_path = ".", boundary = "system") {
    this.default_path = default_path;
    this.boundary = boundary;
    this._handlers = {
      health: (p) => this._m_health(p),
      search: (p) => this._m_search(p),
      symbols: (p) => this._m_symbols(p),
      def: (p) => this._m_def(p),
      ref: (p) => this._m_ref(p),
      read: (p) => this._m_read(p),
      context: (p) => this._m_context(p),
      tree: (p) => this._m_tree(p),
    };
  }



  handle(req) {
    const rid = req.id === undefined ? null : req.id;
    const method = req.method ?? "";


    let raw_params = req.params;
    let params;
    if (!raw_params || (Array.isArray(raw_params) && raw_params.length === 0)) {
      params = {};
    } else if (typeof raw_params !== "object" || Array.isArray(raw_params)) {
      return { id: rid, ok: false, error: "'params' must be an object" };
    } else {
      params = raw_params;
    }

    const handler = Object.prototype.hasOwnProperty.call(this._handlers, method)
      ? this._handlers[method]
      : null;
    if (handler === null) {
      return {
        id: rid, ok: false,
        error: `Unknown method '${method}'; available: ${METHODS.join(", ")}`,
      };
    }
    try {
      const data = handler(params);
      return { id: rid, ok: true, data };
    } catch (e) {

      return { id: rid, ok: false, error: `${e.name}: ${e.message}` };
    }
  }



  _path(params) {
    return params.path || this.default_path;
  }

  _checkPath(params) {
    




    const p = this._path(params);
    if (this.boundary === "root") {
      const base = abs_path(this.default_path);
      const root = abs_path(find_project_root(p));
      if (!is_relative_to(root, base)) {
        throw new PyValueError(`Path '${p}' escapes project root`);
      }
    }
    return p;
  }

  _m_health() {
    return {
      version: VERSION,
      methods: METHODS.slice(),
      root: abs_path(this.default_path),
      boundary: this.boundary,
    };
  }

  _m_search(params) {
    const pattern = params.pattern ?? "";
    if (!pattern || !String(pattern).trim()) {
      throw new PyValueError("Missing 'pattern'");
    }
    let exts = null;
    const raw_exts = params.extensions;
    const norm_ext = (e) => (e.startsWith(".") ? e.toLowerCase() : `.${e.toLowerCase()}`);
    if (Array.isArray(raw_exts)) {
      exts = new Set(raw_exts.map((e) => norm_ext(String(e))));
    } else if (typeof raw_exts === "string") {
      exts = new Set(raw_exts.split(",").map((e) => norm_ext(e)));
    }
    const resp = engine.search_text(
      pattern,
      this._checkPath(params),
      exts,
      params.context ?? 2,
      Boolean(params.ignore_case),
      params.limit ?? 50,
      Boolean(params.whole_word),
    );
    if (!resp.ok) throw new PyValueError(resp.error);
    return engine.search_response_to_dict(resp).data;
  }

  _m_symbols(params) {
    const name = params.name ?? "";
    if (!name) throw new PyValueError("Missing 'name'");
    const resp = engine.search_symbols(
      name,
      this._checkPath(params),
      params.kind ?? null,
      params.limit ?? 50,
      null,
      params.partial === undefined ? true : Boolean(params.partial),
    );
    return engine.symbol_response_to_dict(resp).data;
  }

  _m_def(params) {
    const name = params.name ?? "";
    if (!name) throw new PyValueError("Missing 'name'");
    const resp = engine.find_definition(
      name,
      this._checkPath(params),
      null,
      params.limit ?? 50,
      params.substring_fallback === undefined ? true : Boolean(params.substring_fallback),
    );
    return engine.symbol_response_to_dict(resp).data;
  }

  _m_ref(params) {
    const name = params.name ?? "";
    if (!name) throw new PyValueError("Missing 'name'");
    const resp = engine.find_references(
      name,
      this._checkPath(params),
      params.limit ?? 100,
    );
    if (!resp.ok) throw new PyValueError(resp.error);
    return engine.search_response_to_dict(resp).data;
  }

  _m_read(params) {
    const file_ref = params.file ?? "";
    if (!file_ref) throw new PyValueError("Missing 'file'");
    const result = reader.read_file(
      file_ref,
      this._path(params),
      Boolean(params.outline),
      this.boundary,
    );
    if (!result.ok) throw new PyValueError(result.error);
    return reader.read_result_to_dict(result).data;
  }

  _m_context(params) {
    const file_ref = params.file ?? "";
    const line = safe_int(params.line ?? 0, 0, 0, 1e9);
    if (!file_ref || line <= 0) {
      throw new PyValueError("Missing 'file' and 'line'");
    }
    const result = reader.get_context(
      file_ref,
      line,
      this._path(params),
      safe_int(params.radius ?? 5, 5, 0, 200),
      this.boundary,
    );
    if (!result.ok) throw new PyValueError(result.error);
    return reader.context_result_to_dict(result).data;
  }

  _m_tree(params) {
    return engine.project_tree(
      this._checkPath(params),
      params.depth ?? 3,
    );
  }
}


export function run_rpc(root = null) {
  



  const default_path = root ? root : ".";
  const boundary = root ? "root" : "system";
  const session = new RpcSession(default_path, boundary);


  process.stdout.on("error", (e) => {
    if (e.code === "EPIPE") process.exit(0);
    throw e;
  });




  const tooLarge = {
    id: null, ok: false,
    error: `Request line too large (>${MAX_LINE_BYTES} bytes)`,
  };

  const handleLine = (line) => {
    let resp;
    try {
      const req = JSON.parse(line);
      if (req === null || typeof req !== "object" || Array.isArray(req)) {
        throw new PyValueError("Request must be a JSON object");
      }
      resp = session.handle(req);
    } catch (e) {
      resp = { id: null, ok: false, error: `Invalid request: ${e.message}` };
    }
    process.stdout.write(JSON.stringify(resp) + "\n");
  };

  let buf = Buffer.alloc(0);
  let dropping = false;
  process.stdin.on("data", (chunk) => {
    buf = Buffer.concat([buf, typeof chunk === "string" ? Buffer.from(chunk, "utf8") : chunk]);
    for (;;) {
      const nl = buf.indexOf(0x0a);
      if (nl < 0) {
        if (!dropping && buf.length > MAX_LINE_BYTES) {
          dropping = true;
          buf = Buffer.alloc(0);
          process.stdout.write(JSON.stringify(tooLarge) + "\n");
        }
        return;
      }
      const raw = buf.subarray(0, nl);
      buf = buf.subarray(nl + 1);
      if (dropping) {
        dropping = false;
        continue;
      }
      if (raw.length > MAX_LINE_BYTES) {
        process.stdout.write(JSON.stringify(tooLarge) + "\n");
        continue;
      }

      const line = raw.toString("utf8").trim().replace(/^\uFEFF/, "").trim();
      if (!line) continue;
      handleLine(line);
    }
  });


  process.stdin.on("end", () => {
    if (buf.length && !dropping) {
      if (buf.length > MAX_LINE_BYTES) {
        process.stdout.write(JSON.stringify(tooLarge) + "\n");
      } else {
        const line = buf.toString("utf8").trim().replace(/^\uFEFF/, "").trim();
        if (line) handleLine(line);
      }
    }
    process.exit(0);
  });
}
